src/isolation/landlock.rs

```diff
@@ -0,0 +1,356 @@
+					}),
+			)?
+			.set_compatibility(self.compat_level)
+			.add_rules(
+				self.ro_paths
+					.as_slice()
+					.iter()
+					.map::<Result<_, RestrictError>, _>(|p| {
+						debug!("Adding read-only path ruleset for {:#?}", *p);
+						Self::determine_ruleset(PathFd::new(p)?, abi)
+					}),
+			)?
+			.set_compatibility(self.compat_level)
+			.add_rules(
+				self.ro_dirs
+					.as_slice()
+					.iter()
+					.map::<Result<_, RestrictError>, _>(|p| {
+						debug!("Adding read-only directory ruleset for {:#?}", *p);
+						Ok(PathBeneath::new(PathFd::new(p)?, access_dir_with_rm))
+					}),
+			)?
+			.set_compatibility(self.compat_level)
+			.set_no_new_privs(true)
+			.restrict_self()?;
+
+		debug!(
+			"Landlock ruleset status: {:#?} (no_new_privs: {:#?})",
+			res_status.ruleset, res_status.no_new_privs
+		);
+
+		// Adapted from Mickaël Salaün's Rust Landlock library, which is distributed
+		// under the terms of the MIT and Apache 2.0 licenses.
+		//
+		// For the original source code authored by Mickaël Salaün, see:
+		// https://github.com/landlock-lsm/rust-landlock/blob/07f2a9a7/examples/sandboxer.rs#L133C1-L135C6
+		if res_status.ruleset == RulesetStatus::NotEnforced {
+			panic!(
+				"Landlock is not supported by the running kernel. You can disable Landlock using `--file-isolation none`."
+			);
+		}
+
+		let mut is_already_enabled = LANDLOCK_ENABLED.lock().unwrap();
+		*is_already_enabled = true;
+
+		Ok(res_status)
+	}
+
+	/// Derive a suitable Landlock policy for a given file descriptor.
+	///
+	/// Although applying directory-specific access rights to a file will just result in those rights being ignored
+	/// in "best effort mode", this is not the case when running Uhyve
+	/// in strict mode. Therefore, this function omits directory-specific
+	/// access rights to "R/W" files.
+	///
+	/// - `fd` - File descriptor (of type [`PathFd`])
+	/// - `abi` - Landlock ABI version
+	fn determine_ruleset(fd: PathFd, abi: ABI) -> Result<PathBeneath<PathFd>, RestrictError> {
+		let is_file = Self::is_file(fd.as_fd().as_raw_fd()).unwrap();
+		match is_file {
+			true => Ok(PathBeneath::new(fd, AccessFs::from_file(abi))),
+			false => Ok(PathBeneath::new(fd, AccessFs::from_all(abi))),
+		}
+	}
+
+	/// Checks whether a raw file descriptor represents a file using [`libc::fstat`].
+	///
+	/// Adapted from Mickaël Salaün's Rust Landlock library, which is distributed
+	/// under the terms of the MIT and Apache 2.0 licenses.
+	///
+	/// For the original source code authored by Mickaël Salaün, see:
+	/// https://github.com/landlock-lsm/rust-landlock/blob/aaccff53/src/fs.rs#L198-L210
+	///
+	/// * `rawfd` - Raw file descriptor ([`RawFd`])
+	fn is_file(rawfd: RawFd) -> Result<bool, Error> {
+		unsafe {
+			let mut stat = std::mem::zeroed();
+			match libc::fstat(rawfd, &mut stat) {
+				0 => Ok((stat.st_mode & libc::S_IFMT) != libc::S_IFDIR),
+				// Should not be practically reachable.
+				_ => panic!("stat against fd {:?} failed", rawfd),
```

Comment:
```suggestion
				_ => unreachable!("stat against fd {:?} failed", rawfd),
```
?

Reply:
* under optimal circumstances, correcting the comment, the libc call can just fail, but we hope it shouldn't